### PR TITLE
[feature](restore) support atomic restore

### PR DESCRIPTION
**Pull request by w41ter**
Previously, during a restore job, if the OLAP table already existed, it would be marked with a RESTORE status, preventing any read or write operations.

This PR introduces an atomic restore feature. If enabled, data will first be loaded into a temporary table during the restore, and then atomically replace the original table, ensuring that the target table's read and write operations remain unaffected during the restore process.

### CODE REVIEW

**Comment by w41ter** [comment]
run buildall

**Comment by w41ter** [comment]
run buildall

**Comment by w41ter** [comment]
run buildall

**Comment by w41ter** [comment]
run buildall

**Comment by w41ter** [comment]
run buildall

**Comment by w41ter** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java [hunk 1]
```diff
@@ -2135,6 +2264,86 @@ public void cancelInternal(boolean isReplay) {
...
+        assert isAtomicRestore;
+        for (String tableName : jobInfo.backupOlapTableObjects.keySet()) {
+            String originName = jobInfo.getAliasByOriginNameIfSet(tableName);
+            if (Env.isStoredTableNamesLowerCase()) {
+                originName = originName.toLowerCase();
+            }
+            String aliasName = tableAliasWithAtomicRestore(originName);
+
+            if (!db.writeLockIfExist()) {
+                return Status.OK;
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:2277]: Should we return an error?

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java [hunk 2]
```diff
@@ -2135,6 +2264,86 @@ public void cancelInternal(boolean isReplay) {
...
+                    newOlapTbl.writeUnlock();
+                }
+
+                if (originOlapTbl != null) {
+                    // The origin table is not used anymore, need to drop all its tablets.
+                    originOlapTbl.writeLock();
+                    try {
+                        LOG.info("drop the origin olap table {} by atomic restore. table={}",
+                                originOlapTbl.getName(), originOlapTbl.getId());
+                        Env.getCurrentEnv().onEraseOlapTable(originOlapTbl, isReplay);
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:2334]: For safety, should we put it in recylebin?

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java (hunk 1 above)
Comment (w41ter) [reply fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:2277]: No, the rest of the code does the same thing, so it's best to be consistent here.

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java (hunk 2 above)
Comment (w41ter) [reply fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:2334]: no

**Comment by w41ter** [comment]
run buildall

**Comment by w41ter** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Review by gavinchou — APPROVED** [review]

**Pull request merged**